script/tool/test/mocks.dart

```diff
@@ -32,36 +33,49 @@ class MockPlatform extends Mock implements Platform {
 }
 
 class MockProcess extends Mock implements io.Process {
-  MockProcess();
-
-  /// A mock process that terminates with exitCode 0.
-  MockProcess.succeeding() {
-    exitCodeCompleter.complete(0);
+  /// Creates a mock process with the given results.
+  ///
+  /// The default encodings match the ProcessRunner defaults; mocks for
+  /// processes run with a different encoding will need to be created with
+  /// the matching encoding.
+  MockProcess({
+    int exitCode = 0,
+    String? stdout,
+    String? stderr,
+    Encoding stdoutEncoding = io.systemEncoding,
+    Encoding stderrEncoding = io.systemEncoding,
+  }) : _exitCode = exitCode {
+    if (stdout != null) {
+      _stdoutController.add(stdoutEncoding.encoder.convert(stdout));
+    }
+    if (stderr != null) {
+      _stderrController.add(stderrEncoding.encoder.convert(stderr));
+    }
+    _stdoutController.close();
+    _stderrController.close();
   }
 
   /// A mock process that terminates with exitCode 1.
-  MockProcess.failing() {
-    exitCodeCompleter.complete(1);
-  }
+  MockProcess.failing() : this(exitCode: 1);
```

Comment:
nit: I would inline this function and get rid of it.  It seems to suggest output and exitCode != 0 are mutually exclusive.  What you did with the MockProcess constructor is better.

Reply:
Done. I liked it because it made it clear we didn't care what the exit code was, just that it was non-zero, but as you point out the inconsistency with failing-with-output isn't great (although the latter is comparatively very rare).